File: backend/app/schemas/predictive_maintenance.py
"""预测性维护 Schema — Story 36.3 + 36.4"""

from datetime import datetime
from typing import Literal
from pydantic import BaseModel, ConfigDict, Field


class MaintenanceAdviceInfo(BaseModel):
    """维护建议信息"""

    id: int
    device_id: int
    device_name: str | None = None
    device_type: str | None = None
    health_score: float | None = None
    urgency: Literal["high", "medium"] | None = None
    reason: str | None = None
    suggested_action: str | None = None
    status: Literal["pending", "converted", "rejected", "auto_closed"]
    feedback: str | None = None
    work_order_id: int | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    confirmed_at: datetime | None = None
    confirmed_by: int | None = None
    model_config = ConfigDict(from_attributes=True)


class AdviceRejectRequest(BaseModel):
    """拒绝建议请求"""

    feedback: str = Field(..., min_length=2, max_length=500)


class AdviceConfirmResponse(BaseModel):
    """确认建议响应"""

    advice_id: int
    work_order_id: int
    work_order_no: str
    status: str = "converted"


# ==================== Story 36.4: Dashboard Schema ====================


class DashboardSummary(BaseModel):
    """仪表盘统计概览"""

    total: int
    healthy: int  # 健康
    attention: int  # 关注
    warning: int  # 预警
    danger: int  # 危险


class DeviceHealthItem(BaseModel):
    """设备健康度列表项"""

    device_id: int
    device_name: str | None = None
    device_type: str | None = None
    score: float
    health_level: Literal["健康", "关注", "预警", "危险"]
    data_sufficiency: str | None = None
    degradation_score: float | None = None
    alarm_count: int = 0
    calculated_at: datetime | None = None
    model_config = ConfigDict(from_attributes=True)


class DashboardResponse(BaseModel):
    """仪表盘响应"""

    summary: DashboardSummary
    devices: list[DeviceHealthItem]


class ScoreFactorDetail(BaseModel):
    """评分因子明细"""

    degradation: dict | None = None
    alarm: dict | None = None
    maintenance: dict | None = None
    data_sufficiency: str | None = None
    plugin_key: str | None = None


class DeviceDetailResponse(BaseModel):
    """设备健康度详情响应"""

    health: DeviceHealthItem
    factors: ScoreFactorDetail | None = None
    advices: list[MaintenanceAdviceInfo] = []

File: frontend/src/api/modules/predictiveMaintenance.ts
/**
 * 预测性维护 API 模块 — Story 36.4
 */
import request from '@/utils/request'

// ==================== 类型定义 ====================

export interface DashboardSummary {
  total: number
  healthy: number
  attention: number
  warning: number
  danger: number
}

export interface DeviceHealthItem {
  device_id: number
  device_name: string | null
  device_type: string | null
  score: number
  health_level: '健康' | '关注' | '预警' | '危险'
  data_sufficiency: 'full' | 'partial' | 'minimal' | null
  degradation_score: number | null
  alarm_count: number
  calculated_at: string | null
}

export interface DashboardResponse {
  summary: DashboardSummary
  devices: DeviceHealthItem[]
}

export interface MaintenanceAdviceInfo {
  id: number
  device_id: number
  device_name: string | null
  device_type: string | null
  health_score: number | null
  urgency: 'high' | 'medium' | null
  reason: string | null
  suggested_action: string | null
  status: 'pending' | 'converted' | 'rejected' | 'auto_closed'
  feedback: string | null
  work_order_id: number | null
  created_at: string | null
  updated_at: string | null
  confirmed_at: string | null
  confirmed_by: number | null
}

export interface DeviceDetailResponse {
  health: DeviceHealthItem
  factors: {
    degradation?: { score: number; weight: number }
    alarm?: { score: number; weight: number; count: number }
    maintenance?: { score: number; weight: number; days_since: number | null }
    data_sufficiency?: string
    plugin_key?: string
  } | null
  advices: MaintenanceAdviceInfo[]
}

// ==================== API 函数 ====================

export function getDashboard(params?: {
  device_type?: string
  health_level?: string
  site_id?: number
}) {
  return request.get<DashboardResponse>('/v1/predictive-maintenance/dashboard', { params })
}

export function recalculateHealthScores() {
  return request.post<{
    total_devices: number
    calculated_at: string
    summary: Record<string, number>
    algorithm: string
  }>('/v1/reports/device-health/calculate', undefined, { timeout: 60000 })
}

export function getDeviceDetail(deviceId: number) {
  return request.get<DeviceDetailResponse>(`/v1/predictive-maintenance/devices/${deviceId}/detail`)
}

export function getAdviceList(params?: {
  status?: string
  device_type?: string
}) {
  return request.get<MaintenanceAdviceInfo[]>('/v1/predictive-maintenance/advices', { params })
}

export function confirmAdvice(adviceId: number) {
  return request.post<{ advice_id: number; work_order_id: number; work_order_no: string; status: string }>(
    `/v1/predictive-maintenance/advices/${adviceId}/confirm`
  )
}

export function rejectAdvice(adviceId: number, feedback: string) {
  return request.post<MaintenanceAdviceInfo>(
    `/v1/predictive-maintenance/advices/${adviceId}/reject`,
    { feedback }
  )
}
